### PR TITLE
Fixed #34691 -- Added system check for unmatched angle brackets in path().

### PR DESCRIPTION
**Pull request by AMK9978**
I added the bracket checker like what we have for $. This is my first contribution to the Django project! Let me know if there is anything to modify :)

I'm doubtful of naming the warning id (I know it might happen by upgrading Django, but I don't know in which version it will be used). Do I need to update it to 5? or what? Plus, do I need to include it in the docs/releases/4.3.txt? or docs/releases/5.0.txt?

### CODE REVIEW

File: docs/ref/checks.txt [hunk 1]
```diff
@@ -118,6 +118,8 @@ upgrading Django.
 * **4_0.E001**: As of Django 4.0, the values in the
   :setting:`CSRF_TRUSTED_ORIGINS` setting must start with a scheme (usually
   ``http://`` or ``https://``) but found ``<hostname>``.
+* **4_3.W001**: Your URL pattern ``<pattern>`` has a ``route`` that contains
+  ``<`` or ``>``, begins with a ``<`` or ends with a ``>``, but is not matched.
```
Comment (felixxm) [inline docs/ref/checks.txt:122]: First of all, we don't work on Django 4.3, but on Django 5.0. Secondly, it's not a warning about potential issue after upgrading Django. I'd move it to `urls`.

File: docs/ref/checks.txt [hunk 2]
```diff
@@ -609,6 +611,8 @@ The following checks are performed on your URL configuration:
   imported.
 * **urls.E009**: Your URL pattern ``<pattern>`` has an invalid view, pass
   ``<view>.as_view()`` instead of ``<view>``.
+* **urls.W010**: Your URL pattern ``<pattern>`` has an invalid view, pass
+  ``<view>.as_view()`` instead of ``<view>``.
```
Comment (felixxm) [inline docs/ref/checks.txt:615]: Please revert.

File: tests/check_framework/urls/path_compatibility/ending_with_angle_bracket.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from django.urls import path
+
+urlpatterns = [
+    path("ending-with/angle-bracket>", lambda x: x),
+]
```
Comment (felixxm) [inline tests/check_framework/urls/path_compatibility/ending_with_angle_bracket.py:5]: I'd put both incorrect patterns in a single file.

**Comment by AMK9978** [comment]
@felixxm 
I just resolved all of your comments. Thnx for your notes :)

File: tests/check_framework/test_urls.py [hunk 4]
```diff
@@ -196,6 +196,43 @@ def test_ending_with_dollar(self):
...
+        ROOT_URLCONF="check_framework.urls.path_compatibility.matched_angle_brackets"
+    )
+    def test_matching_angle_brackets(self):
+        result = check_url_config(None)
+        self.assertEqual(len(result), 0)
+
+    @override_settings(
+        ROOT_URLCONF="check_framework.urls.path_compatibility.mismatched_angle_brackets"
+    )
+    def test_get_warning_beginning_with_angle_bracket(self):
```
Comment (felixxm) [inline tests/check_framework/test_urls.py:209]: We can do this in a single test:
```suggestion
    def test_unmatched_brackets(self):
```
For example:
```python
    @override_settings(
        ROOT_URLCONF="check_framework.urls.path_compatibility.mismatched_angle_brackets"
    )
    def test_unmatched_brackets(self):
        result = check_url_config(None)
        expected_msg = (
            "Your URL pattern pattern %r has a route that contains `<` or '>', begins "
            "with a `<` or ends with a '>', but is not matched."
        )
        self.assertEqual(
            result,
            [
                Warning(
                    expected_msg % "beginning-with/<angle-bracket",
                    id="urls.W006",
                ),
                Warning(
                    expected_msg % "ending-with/angle-bracket>",
                    id="urls.W006",
                ),
            ]
        )
```
Also, this test should be moved to the `CheckUrlConfigTests` class.

File: tests/check_framework/urls/path_compatibility/mismatched_angle_brackets.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+from django.urls import path
```
Comment (felixxm) [inline tests/check_framework/urls/path_compatibility/mismatched_angle_brackets.py:1]: Rename file to `unmatched_angle_brackets.py`.

File: tests/check_framework/urls/path_compatibility/matched_angle_brackets.py [hunk 6]
```diff
@@ -0,0 +1,5 @@
+from django.urls import path
```
Comment (felixxm) [inline tests/check_framework/urls/path_compatibility/matched_angle_brackets.py:1]: Rename file to `matched_angle_brackets.py`.

File: tests/check_framework/urls/path_compatibility/mismatched_angle_brackets.py [hunk 7]
```diff
@@ -0,0 +1,6 @@
...
+
+urlpatterns = [
+    path("beginning-with/<angle-bracket", lambda x: x),
+    path("ending-with/angle-bracket>", lambda x: x),
```
Comment (felixxm) [inline tests/check_framework/urls/path_compatibility/mismatched_angle_brackets.py:5]:
```suggestion
    path("beginning-with/<angle_bracket", lambda x: x),
    path("ending-with/angle_bracket>", lambda x: x),
```

File: django/urls/resolvers.py [hunk 8]
```diff
@@ -329,8 +334,22 @@ def check(self):
                     id="2_0.W001",
                 )
             )
+        warnings.extend(self._check_angle_bracket_pattern())
         return warnings
 
+    def _check_angle_bracket_pattern(self):
+        match = re.match(_PATH_ANGLE_BRACKET_COMPONENT_RE, str(self._route))
```
Comment (felixxm) [inline django/urls/resolvers.py:341]: Do we need a regular expression? :thinking: Is is not enough to count `<` and `>`.

File: tests/check_framework/urls/path_compatibility/matched_angle_brackets.py (hunk 6 above)
Comment (AMK9978) [reply tests/check_framework/urls/path_compatibility/matched_angle_brackets.py:1]: It's already is

File: django/urls/resolvers.py (hunk 8 above)
Comment (AMK9978) [reply django/urls/resolvers.py:341]: I think it's a solid way for unmatched angle brackets in general. For example:
'closed_angle>/x/<open_angled'

Comment (felixxm) [reply django/urls/resolvers.py:341]: Good point :+1:

**Comment by felixxm** [comment]
@adamchainz Do you want to take a look?

File: django/urls/resolvers.py [hunk 9]
```diff
@@ -329,8 +334,22 @@ def check(self):
                     id="2_0.W001",
                 )
             )
+        warnings.extend(self._check_angle_bracket_pattern())
```
Comment (felixxm) [inline django/urls/resolvers.py:337]: You can move it to the first line, e.g.
```python
def check(self):
    warnings = [
        *self._check_pattern_startswith_slash(),
        *self._check_angle_bracket_pattern(),
    ]
    ...
```

File: django/urls/resolvers.py [hunk 10]
```diff
@@ -329,8 +334,22 @@ def check(self):
...
         return warnings
 
+    def _check_angle_bracket_pattern(self):
```
Comment (felixxm) [inline django/urls/resolvers.py:340]: IMO, `_check_pattern_unmatched_angle_bracket()` would be more consistent:
```suggestion
    def _check_pattern_unmatched_angle_bracket(self):
```

File: django/urls/resolvers.py [hunk 11]
```diff
@@ -329,8 +334,22 @@ def check(self):
...
+        match = re.match(_PATH_ANGLE_BRACKET_COMPONENT_RE, str(self._route))
+        if not match:
+            warning = Warning(
+                "Your URL pattern pattern {} has a route that contains `<` or '>', "
+                "begins with a `<` or ends with a '>', but is not matched.".format(
```
Comment (felixxm) [inline django/urls/resolvers.py:345]: Maybe: _"Your URL pattern pattern %s has a route that contains unmatched angle brackets '<' or '>'. This is likely a typo."_.

**Review by felixxm — COMMENTED** [review]
@AMK9978 Thanks :+1: I left some suggestions.

File: django/urls/resolvers.py (hunk 11 above)
Comment (AMK9978) [reply django/urls/resolvers.py:345]: U're right!

File: django/urls/resolvers.py (hunk 9 above)
Comment (AMK9978) [reply django/urls/resolvers.py:337]: Nice! I think we can even refactor this check:
`        if "(?P<" in route or route.startswith("^") or route.endswith("$"):
            warnings.append(
                Warning(
                    "Your URL pattern {} has a route that contains '(?P<', begins "
                    "with a '^', or ends with a '$'. This was likely an oversight "
                    "when migrating to django.urls.path().".format(self.describe()),
                    id="2_0.W001",
                )
            )`

as a separate method like two others, and then just something like:
`
        warnings = [
            *self._check_pattern_startswith_slash(),
            *self._check_pattern_unmatched_angle_bracket(),
             self._check_pattern_caret_and_dollar()
        ]`

Then, it might look more consistent and more readable.

Comment (felixxm) [reply django/urls/resolvers.py:337]: > I think we can even refactor this check

I'm not sure it's worth doing. It's not related with this patch so don't do this here.

Comment (AMK9978) [reply django/urls/resolvers.py:337]: Yes, I won't include it in this patch. Maybe can be discussed as a ticket. Probably, such things have appeared more than once in the project

Comment (felixxm) [reply django/urls/resolvers.py]: > Maybe can be discussed as a ticket.

Ticket is not required for small cleanups. In general, we try to avoid trivial code reformatting unless it's a big problem as it creates extra noise in the history.

Comment (AMK9978) [reply django/urls/resolvers.py]: Aha! Got it. I try to get aligned!

**Comment by AMK9978** [comment]
@felixxm 
It seems we have some errors in Github actions that are out of the scope of this patch. I just wanted to make sure everything is okay for the merge.

File: django/urls/resolvers.py [hunk 12]
```diff
@@ -293,6 +293,11 @@ def _route_to_regex(route, is_endpoint=False):
     return "".join(parts), converters
 
 
+_PATH_ANGLE_BRACKET_COMPONENT_RE = _lazy_re_compile(
+    r"^(?:\/?[^<>\/]*(?:<(?!\/)[^<>\/]*>[^<>\/]*)*)*\/?$"
+)
```
Comment (adamchainz) [inline django/urls/resolvers.py:298]: This regular expression is way too complicated to understand, verify, and maintain. Can we go with a much easier approach, please? I suggest we find all the angle brackets, and check for unmatched ones with a loop like so:

```python
import re

path = "users/<int:user_id>/comments/<int:comment_id"

brackets = re.findall("[<>]", path)
opened = False
unmatched = False

for b in brackets:
    if b == "<":
        if not opened:
            opened = True
        else:
            unmatched = True
            break
    else:  # >
        if opened:
            opened = False
        else:
            unmatched = True
            break
if opened:
    unmatched = True

if unmatched:
    # errors.append(...)
```

I haven't tested every combination here, please do test it.

File: django/urls/resolvers.py [hunk 13]
```diff
@@ -331,6 +339,19 @@ def check(self):
             )
         return warnings
 
+    def _check_pattern_unmatched_angle_brackets(self):
+        match = re.match(_PATH_ANGLE_BRACKET_COMPONENT_RE, str(self._route))
+        if not match:
+            warning = Warning(
+                "Your URL pattern {} has a route that contains `<` or '>', "
+                "begins with a `<` or ends with a '>', but is not matched.".format(
+                    self.describe()
```
Comment (adamchainz) [inline django/urls/resolvers.py:348]: This message is too wordy. I suggest:

> URL pattern {} has an unmatched {} bracket

We can adjust to include *which* kind of bracket is unmatched in the second format parameter there. This can be determined from the `opened` variable that I suggested above. If `opened` is true, `<` is unmatched, otherwise `>`.

File: django/urls/resolvers.py (hunk 12 above)
Comment (AMK9978) [reply django/urls/resolvers.py:298]: My first implementation was with using Stack and it will simply match every '<' with a '>' in each segment, otherwise raises a warning. I thought it might be more consistent with the rest of our code, which is using regex.
For the sake of understandability, I can add a comment above that regex explaining it.

Comment (adamchainz) [reply django/urls/resolvers.py:298]: Please switch to the stack-based approach. Consistency does not trump comprehensibility.

Comment (AMK9978) [reply django/urls/resolvers.py]: I switched. Let me know if it still looks complicated. Plus, I could've used a **counter** instead of a stack, and instead of appending to the stack, make the `counter += 1,` and instead of popping from the stack, make the `counter -=1`. If the `counter < 0`, means it has an extra `'>'`, if the `counter > 0`: means there is an extra `'<'`. 
I wasn't sure if it is more readable than the current stack-based way. If you feel the counter is eloquent enough, I can switch to it.

Comment (felixxm) [reply django/urls/resolvers.py]: For me, a version with counter is readable enough. Updated.

File: django/urls/resolvers.py [hunk 14]
```diff
@@ -331,6 +334,34 @@ def check(self):
             )
         return warnings
 
+    def _check_pattern_unmatched_angle_brackets(self):
+        warnings = []
+        segments = self._route.split("/")
```
Comment (adamchainz) [inline django/urls/resolvers.py:339]: Can we avoid splitting on `/`? From the perspective of this check there's nothing special about this character in the URL.

Comment (felixxm) [reply django/urls/resolvers.py:339]: Unmatched angle brackets cannot cross `/` as it raises `ImproperlyConfigured`: _"... uses parameter name ... which isn't a valid Python identifier."_

Comment (adamchainz) [reply django/urls/resolvers.py:339]: Yes, that's why I suggest we don't split on it, as the `RoutePattern` can't exist in a state where such invalid characters exist between `<` and `>`?

Comment (felixxm) [reply django/urls/resolvers.py:339]: > Damn @felixxm merged.

I regret nothing :wink:

Comment (adamchainz) [reply django/urls/resolvers.py:339]: IMG_4777
